tsdate/core.py

```diff
@@ -523,6 +524,78 @@ def scale_geometric(self, fraction, value):
         return fraction * value
 
 
+class VariationalLikelihoods:
+    """
+    A class to store and process likelihoods for use in variational inference.
+    """
+
+    probability_space = base.PAR
+    identity_constant = np.array([1.0, 0.0], dtype=float)  # "improper" gamma prior
+    null_constant = 0.0
+    timepoints = np.array([0, np.inf], dtype=float)
+    grid_size = 2
+
+    def __init__(
+        self,
+        ts,
+        mutation_rate=None,
+        recombination_rate=None,
+        *,
+        fixed_node_set=None,
+    ):
+        self.ts = ts
+        self.fixednodes = (
+            set(ts.samples()) if fixed_node_set is None else fixed_node_set
+        )
+        self.mut_rate = mutation_rate
+        self.rec_rate = recombination_rate
+        self.mut_edges = self.get_mut_edges(ts)
+        self.identity_constant.flags.writeable = False
+        self.timepoints.flags.writeable = False
+
+    def to_gamma(self, edge):
+        """
+        Return the shape and rate parameters of the (gamma) posterior of edge
+        length, given an improper (constant) prior.
+        """
+        y = self.mut_edges[edge.id]
+        mu = edge.span * self.mut_rate
+        return np.array([y + 1, mu])
+
+    @staticmethod
+    def get_mut_edges(ts):
+        """
+        Get the number of mutations on each edge in the tree sequence.
+        """
+        return Likelihoods.get_mut_edges(ts)
+
+    @staticmethod
+    def combine(base, message):
+        """
+        Multiply two gamma PDFs
```

Comment:
I'm not 100% sure how this works (i.e. where the [-1,0] comes from). We should probably state what format the "base" and "message" parameters are (i.e. a pair of numbers, representing the alpha and beta of the gamma distribution, I think?)

Reply:
The [-1, 0] is the conversion between the typical gamma parameterization and it's natural (exponential-family) parameterization. Can definitely add more description